Read Only profile retirement

capeterson opened the pull request:
Update ESAPI tests to use the Minimum Access - Salesforce profile, and add a tiny amount of new debug output for FLS-driven access failures.

capeterson committed:
Add WARN level debug messages when FLS causes an Object level access check to fail

This helps with production debugging - but doesn't show the extra info to the end user, only to an admin who enables and views debug logs.

capeterson committed:
Updated test cases to use Minimum Access profile instead of Read Only

As the Read Only profile is going away soon-ish.

capeterson commented:
@FishOfPrey since you wrote a big chunk of this code, game to review it before we merge it?

In prep for https://help.salesforce.com/articleView?id=000355656&type=1&mode=1, where the Read Only profile will no longer reliably exist.

FishOfPrey commented:
From a reading of the diff on the Minimum Access profile change it seems like a good change. I'm assuming that Email and ReportsTo needed to be added to the default read/write fields as the historical Read Only profile granted them by default?

 The additional debug log warnings are an interesting one. I agree that they are definitely useful when debugging. My concern would be the performance impact from building up the string messages, even if warning level logging isn't on. The rest of the class is devoid of logging for that reason.
 
 I'll try and test this pull request tomorrow.

capeterson commented:
Yup - I swapped the test creation over to minimum access, got 3 failures that confused me deeply since I saw we were creating object level access. That's when I added the debug statements, since finding out WHAT field level access blocked you was.... challenging, and I can imagine users have the same issue with production debugging of ESAPI errors.

Those fields were what was causing it to fail, since they were on the list to validate, but not on the list the helper method created access for, so I added them to compensate for the lack of rights on Minimum Access that ReadOnly had. 

>My concern would be the performance impact from building up the string messages, even if warning level logging isn't on. The rest of the class is devoid of logging for that reason.

Considering it's within an inner loop of describe calls that are going to be MASSIVELY slower, and only runs once per rejection, and only if it was rejected due to FLS, I feel pretty comfortable with the impact here. If we were logging on each loop iteration I'd agree, but once for a DML option we kick back is... well, a lot cheaper than the DML operation going through would be the crux of my argument.
